### ISSUE TITLE
Question on supervisor function

### ISSUE THREAD

**Issue by karkawal** [jsyoon0823/TimeGAN#4]
I am struggling to understand the actual function of `supervisor`.
`G_loss_S = tf.losses.mean_squared_error(H[:,1:,:], H_hat_supervise[:,1:,:])`
What this line acutally does? Doesn't it force `supervisor` to learn identity between H and H_hat_supervise?

What sequence should the supervisor generate? Is it generating from [0, T] to [1, T+1] or is it trying to reflect the original sequence [0, T]?

**Comment by jsyoon0823** [comment]
Thank you for your question.

There is a typo on the supervised loss and it should be as follows.
G_loss_S = tf.losses.mean_squared_error(H[:,1:,:], H_hat_supervise[:,:-1,:])
The main objective of the supervisor is "Generate next sequence using the previous sequence".
Therefore, G_loss_S forces the supervisor to generate next sequence (H[:,1:,:]) using previous sequence (H[:,:-1,:]).

Thanks!

**Comment by zzhang222** [comment]
Should the var_list of GS_solver be 's_vars' then?

**Comment by zzhang222** [comment]
And you input z when training the supervisor, which seems not necessary.

**Comment by jsyoon0823** [comment]
Actually, generating supervisor output H needs E which is generated by generator.
Therefore, for GS_solver, we need both g_vars and s_vars.
Also, to generate E, we need z. Thus, z_input is also needed.
Thanks!

**Comment by zzhang222** [comment]
I guess you should have H_hat_supervise = supervisor(H_hat, T) instead of H_hat_supervise = supervisor(H, T) on line 171. Since in the original code, the loss is not dependent on E.

**Comment by jsyoon0823** [comment]
Thanks for checking that. In that point of view, it would be fine to remove g_vars from GS_solver.
H_hat_supervise = supervisor(H, T) is intended because we want to make supervisor that follows the original sequence.

**Comment by zzhang222** [comment]
It makes sense now, thank you!
